BIP-112: Mempool-only CHECKSEQUENCEVERIFY

btcdrak opened the pull request:
Replace NOP3 with CHECKSEQUENCEVERIFY (BIP-112)

The BIP text is at https://github.com/bitcoin/bips/blob/master/bip-0112.mediawiki

This PR follows on from #6564 and is rebased (with #7184 which has been merged top `master`) so can now be tested directly.

Functional test scripts by @ajtowns can be found at https://github.com/ajtowns/op_csv-test

For reviewers please note that #6564 remained unchanged for several months. The list of reviews in #6564 were as follows:
- Kanzure [utACK](https://github.com/bitcoin/bitcoin/pull/6564#issuecomment-144863501)
- rustyrussell [utACK](https://github.com/bitcoin/bitcoin/pull/6564#issuecomment-147272511)
- instagibbs [utACK](https://github.com/bitcoin/bitcoin/pull/6564#issuecomment-149951293)
- dcousens [utACK](https://github.com/bitcoin/bitcoin/pull/6564#issuecomment-150043472)
- rubensayshi [utACK](https://github.com/bitcoin/bitcoin/pull/6564#issuecomment-150164947)
- afk11 [utACK](https://github.com/bitcoin/bitcoin/pull/6564#issuecomment-150265364)
- jgarzik [utACK](https://github.com/bitcoin/bitcoin/pull/6564#issuecomment-156357583)
- greenaddress [utACK](https://github.com/bitcoin/bitcoin/pull/6564#issuecomment-158037452)
- rustyrussel [Tested ACK](https://github.com/bitcoin/bitcoin/pull/6564#issuecomment-169546151)
- ajtowns [Tested ACK](https://github.com/bitcoin/bitcoin/pull/6564#issuecomment-176257453)
- josephpoon [Tested ACK](https://github.com/bitcoin/bitcoin/pull/6564#issuecomment-178998124)
- CodeShark [Tested ACK](https://github.com/bitcoin/bitcoin/pull/6564#issuecomment-180860229)
- NicolasDorier [Tested ACK](https://github.com/bitcoin/bitcoin/pull/6564#issuecomment-182324514)

maaku committed:
BIP112: Implement CHECKSEQUENCEVERIFY

- Replace NOP3 with CHECKSEQUENCEVERIFY (BIP112)
  <nSequence> CHECKSEQUENCEVERIFY -> <nSequence>
- Fails if txin.nSequence < nSequence, allowing funds of a txout to be locked for a number of blocks or a duration of time after its inclusion in a block.
- Pull most of CheckLockTime() out into VerifyLockTime(), a local function that will be reused for CheckSequence()
- Add bitwise AND operator to CScriptNum
- Enable CHECKSEQUENCEVERIFY as a standard script verify flag
- Transactions that fail CSV verification will be rejected from the mempool, making it easy to test the feature. However blocks containing "invalid" CSV-using transactions will still be accepted; this is *not* the soft-fork required to actually enable CSV for production use.

NicolasDorier commented:
Tested ACK again. I rebased myself maaku's branch and compared with btcdrak, it is correct. For the lazy:

```
git checkout master
git pull origin master

git remote add btcdrak https://github.com/btcdrak/bitcoin
git remote add maaku https://github.com/maaku/bitcoin

git fetch btcdrak checksequenceverify
git fetch maaku checksequenceverify

git checkout -b btcdrakchecksequenceverify btcdrak/checksequenceverify
git checkout -b maakuchecksequenceverify maaku/checksequenceverify

git rebase --onto 80d1f2e48364f05b2cdf44239b3a1faa0277e58e ce15c73964b14a6ede4b713021147060b23a71e8 HEAD
#Solve by taking "Theirs version"
git diff HEAD btcdrakchecksequenceverify
```

NicolasDorier commented:
Tested ACK again, I rebased maaku's branch and diff with btcdrak's e2b7001831cfb75e06334ba7c2cb026f8ed96434, this is correct.
For the lazy:
```
git checkout master
git pull origin master

git remote add btcdrak https://github.com/btcdrak/bitcoin
git remote add maaku https://github.com/maaku/bitcoin

git fetch btcdrak checksequenceverify
git fetch maaku checksequenceverify

git checkout -b btcdrakchecksequenceverify btcdrak/checksequenceverify
git checkout -b maakuchecksequenceverify maaku/checksequenceverify

git rebase --onto 80d1f2e48364f05b2cdf44239b3a1faa0277e58e ce15c73964b14a6ede4b713021147060b23a71e8 HEAD
#Solve by taking "Theirs version"
git diff HEAD btcdrakchecksequenceverify
```

btcdrak committed:
Separate CheckLockTime() and CheckSequence() logic

For the sake of a little repetition, make code more readable.

morcos commented:
My [comment](https://github.com/bitcoin/bitcoin/pull/6564#issuecomment-178636856) remains that I would prefer VerifyLockTime not be encapsulated for reuse for nLockTime and nSequence checks.

NicolasDorier commented:
@morcos @petertodd @btcdrak I rewrote history of this PR to remove any change to CheckLockTime: https://github.com/bitcoin/bitcoin/compare/master...NicolasDorier:btcdrakchecksequenceverify2#diff-be2905e2f5218ecdbe4e55637dac75f3R1196

This make a small diff.

btcdrak commented:
@morcos @petertodd I agree the encapsulation hampers readability. I propose something like 2fcd56f.

@NicolasDorier Thank you, but I am not interested in rebasing history at this stage of the PR.

NicolasDorier commented:
ACK 2fcd56f

instagibbs commented:
utACK https://github.com/btcdrak/bitcoin/commit/2fcd56f70845e9c78ab3ffde38fbb11ccf33e6da

btcdrak commented:
I've squashed to two commits. 53e53a3 corresponds to the state of #6564 as per OP and c3c3752 (was 2fcd56f) addresses request from @morcos and @petertodd.

No change in diff for verification [`checksequenceverify...btcdrak:checksequenceverify_2fcd56f`](https://github.com/btcdrak/bitcoin/compare/checksequenceverify...btcdrak:checksequenceverify_2fcd56f).

laanwj commented on src/script/interpreter.cpp:
```diff
@@ -1156,6 +1193,50 @@ bool TransactionSignatureChecker::CheckLockTime(const CScriptNum& nLockTime) con
+        (txToSequenceMasked <  CTxIn::SEQUENCE_LOCKTIME_TYPE_FLAG && nSequenceMasked <  CTxIn::SEQUENCE_LOCKTIME_TYPE_FLAG) ||
+        (txToSequenceMasked >= CTxIn::SEQUENCE_LOCKTIME_TYPE_FLAG && nSequenceMasked >= CTxIn::SEQUENCE_LOCKTIME_TYPE_FLAG)
+    ))
+        return false;
```
This `return false` here seems to come out of the blue, which could make people misread the code and introduce bugs. I'd suggest adding the `))` to the line before it, or surrounding with `{}`.

laanwj commented:
utACK 2fcd56f

btcdrak committed:
Code style fix.

This if statement is a little obtuse and using braces here
improves readability.
